TheLastScrub/delta-green-foundry-vtt-system-unofficial
# Roll Refactor | Sheet Restyle

This PR does a number of things:

* Refactor roll system
  * Full rebuild of the roll system. 
  * Separate rolls into their own class `DGRoll` which extends base Foundry `Roll`
  * `DGRoll` is a basic/general class that specific rolls should extend. We almost never do `new DGRoll(...args)` directly.
  * For example, `DGPercentileRoll` extends `DGRoll`, and applies its own methods and data to the class object.
  * This has simplified a lot of nested if/else statements and rolls now follow a more understandable and consistent logic.
  * This has also allowed us to remove a lot of clutter in the datasets in html elements.
    * Percentile rolls just need to specify `data-rolltype`, `data-key`, and `data-iid` in the HTML. 
      * The `rolltype` for percentile rolls should be one of the following values `"skill"`, `"stat"`, `"weapon"`, `"lethality"`, `"luck"`, . Note, Lethality rolls are a subclass of percentile rolls, not damage rolls.
      * The `key` should be the stat (e.g. `"str"`, skill (e.g. `"unarmed-combat"`), or `"custom"` that is used to derive the target number. This is unnecessary for `lethality`, `sanity`, `luck`, and damage rolls.
      * The `iid` should be the `ID` of the item used to make the roll, if applicable. Note, this is only necessary to provide when rolling from an item like with a weapon attack or damage/lethality roll. 
    * Damage rolls need to specify `data-rolltype` and `data-iid`
      * The `rolltype` for damage rolls should be `damage` or `sanity-damage`.
      * `data-iid` is the id of the weapon used to make the damage roll. Not necessary for sanity damage rolls
* The refactor of the roll system brought along some slight restyling of roll chat messages.
* Restyled actor sheets to be more in-line with the core book's design.
* Weapons can now have modifiers for their rolls built-in.
* Weapons can choose a custom target for their rolls.
* Damage rolls can now utilize math functions like base Foundry rolls (e.g. `ceil((1d12+2)/4)`).
* The "Hide Untrained Skills" checkbox has been turned into an "eye" glyph.

## Steps to test:

For all rolls, try rolling a few to make sure all the combinations work (critical success, regular success, regular fail, and critical fail).

1. Roll stats
    1. Roll unmodified stat
    2. Roll modified stat
    3. Roll inhuman stat (stat > 20) and make sure it follows all the rules
    4. Check chat cards (expanded and not expanded) and make sure they look good.
2. Roll skills
    1. Roll unmodified skill
    2. Roll unmodified typed skill
    3. Roll modified skill
    4. Roll unmodified typed skill
    5. Check chat cards (expanded and not expanded) and make sure they look good.
3. Roll Luck check, modified and unmodified
4. Roll a Sanity check, modified and unmodified
5. Roll weapons
    1. Roll weapon attack (using a skill for the target) modified and unmodified
    2. Roll regular damage modified and unmodified
    3. Roll regular damage with a math function in the formula
    4. Roll lethality damage modified and unmodified
    5. Repeat steps 5.1-5.4 using a stat x5 for the target
    6. Repeat steps 5.1-5.4 using a custom target
    7. Repeat steps 5.1-5.4 with a modifier on the weapon item. 
    8. Check chat messages for 5.1 - 5.7.
6. Repeats steps 1-5 on NPC and Unnatural sheets (if applicable).
7. On an unnatural sheet, roll sanity damage. Check chat message
8. On NPC/Unnatural sheets, click the "eye" glyph to toggle the display of untrained skills.
